File: vm/js/format/src/instruction-util.ts
import { Chunk } from "./chunk.js";

export interface InstructionContext {
    readonly chunk: Chunk;
    ip: number;
}

export function readIndex(ctx: InstructionContext): bigint {
    let result = 0n;
    let b;
    let shift = 0n;
    
    do {
        const item = ctx.chunk.bytecode[ctx.ip];
        if(item === undefined) {
            throw new Error("Unexpected end of data");
        }
        ++ctx.ip;

        b = item;

        result = (result << shift) | (BigInt(b & 0x7f));
        shift += 7n;
    } while((b & 0x80) != 0);

    return result;
}

export function readInt8(ctx: InstructionContext): number {
    const b = ctx.chunk.bytecode[ctx.ip];
    if(b === undefined) {
        throw new Error("Unexpected end of data");
    }
    ++ctx.ip;

    return b << 24 >> 24;
}

File: vm/rust/format/src/instruction_util.rs
use std::io::{Write, Result, Read};


pub fn write_index<F: Write>(f: &mut F, mut value: usize) -> Result<()> {
    loop {
        let next_part = (value & 0x7F) as u8;
        value = value >> 7;
        if value > 0 {
            f.write(std::slice::from_ref(&(next_part | 0x80)))?;
        }
        else {
            f.write(std::slice::from_ref(&next_part))?;
            return Ok(());
        }

    }
}



pub fn read_index<F: Read>(f: &mut F) -> Result<usize> {
    let mut result: usize = 0;

    loop {
        let mut b: u8 = 0;
        f.read_exact(std::slice::from_mut(&mut b))?;

        result = (result << 7) | (b & 0x7F) as usize;
        if (b & 0x80) == 0 {
            break;
        }
    }

    Ok(result)
}

pub fn write_int8<F: Write>(f: &mut F, value: i8) -> Result<()> {
    let b = value as u8;
    f.write(std::slice::from_ref(&b))?;
    Ok(())
}

pub fn read_int8<F: Read>(f: &mut F) -> Result<i8> {
    let mut b: u8 = 0;
    f.read_exact(std::slice::from_mut(&mut b))?;
    Ok(b as i8)
}
